### DISCUSSION TITLE
Should DROP SCHEMA succeed or fail when there is a table in the schema?

### DISCUSSION THREAD

**Discussion by findepi** [trinodb/trino#8613] [General]
Should DROP SCHEMA succeed or fail when there is a table in the schema?

@martint @kasiafi

**Comment by kokosing** [comment]
It should fail.

Notice that currently DROP SCHEMA contains option CASCADE or RESTRICT.

CASCADE - Automatically drop objects (tables, functions, etc.) that are contained in the schema. Currently we throw an error that it is not yet supported.

RESTRICT - Refuse to drop the schema if it contains any objects. This is the default when no option is given.

**Comment by findepi** [comment]
> RESTRICT - Refuse to drop the schema if it contains any objects. This is the default when no option is given.

Awesome. But seems we do not verify it, do we?

**Reply by martint** [reply]
From the SQL specification:

> If RESTRICT is specified, then S shall not contain any persistent base tables, global temporary tables, created local temporary tables, views, domains, assertions, character sets, collations, transliterations, triggers, user-defined types, SQL-invoked routines, sequence generators, or roles, and the `<schema name>` of S shall not be contained in the SQL routine body of any routine descriptor.
> NOTE 519 — If CASCADE is specified, then such objects will be dropped implicitly by the `<revoke statement>` and/or explicitly by the SQL-schema manipulation statements specified in the General Rules of this Subclause.

**Reply by findepi** [reply]
> From the SQL specification:

@martint  Thanks!

do you know what should be happening when none of RESTRICT nor CASCADE is specified?

**Reply by martint** [reply]
In the spec, the CASCADE or RESTRICT clause is actually required. The behavior in Trino is an extension, and I don't think we've ever defined what should happen if it's omitted.

**Reply by findepi** [reply]
Currently, the code ignores RESTRICT clause, as if assuming that default behavior is RESTRICT.

if it was for me to choose (as spec doesn't define this), RESTRICT by default is the only reasonable default.

we have a bug in MySQL though https://github.com/trinodb/trino/issues/8634 
and we should document all this
cc @mosabua

**Reply by mosabua** [reply]
Not sure what specifically we want to document and where.

Maybe just mention how DROP SCHEMA behaves now? And then adjust whenever we change?

And I assume we would document that in the drop schema page .. do we need details for each connector that supports it? We might if it varies..

**Comment by linzebing** [comment]
Is there any plan to support drop schema cascade?

**Reply by ebyhr** [reply]
https://github.com/trinodb/trino/pull/17855
